### PR TITLE
[Merged by Bors] - feat: port Algebra.Group.WithOne.Defs

### PR DESCRIPTION
**Pull request by kbuzzard**
mathlib3 commit hash e574b1a4e891376b0ef974b926da39e05da12a06

### CODE REVIEW

**Comment by kbuzzard** [comment]
Things of note here:

1) Like `WithBot`, we add `coe : α → WithOne α` and tag it `@[coe]` for compatibility with Lean 4's coercion mechanism
2) Neither `@[coe, to_additive]` nor `@[to_additive, coe]` seems to tag the additivised definition with `coe`; same for `@[norm_cast, to_additive]` (which used to work in Lean 3). I've had to manually tag some to-additivised declarations as a consequence.
3) I had to comment out a `CanLift` instance, and also rewrite a proof, because there is still no `lift` tactic.
4) There are some `show`s which might not be needed; I asked at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/_proof_1/near/313734606 because I don't know how to test this.
5) `WithZero.zero_mul` with `WithZero.mul_zero` used to be tagged `@[simp]` but a linter told me that `simp only [zero_mul]` proves it, so I removed the tags. This makes sense: `WithZero alpha` is an instance of `MulZeroClass`. I'm slightly surprised that the linter in mathlib3 doesn't flag this.

File: Mathlib/Algebra/Group/WithOne/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+  mul_one := show ∀ x : WithOne α, x * 1 = x from (Option.liftOrGet_isRightId _).1
+
+@[to_additive]
+instance [Semigroup α] : Monoid (WithOne α) :=
+  { instMulOneClassWithOne with mul_assoc := (Option.liftOrGet_isAssociative _).1 }
+
+-- example [Semigroup α] :
+--     @Monoid.toMulOneClass _ (@instMonoidWithOne α _) = @instMulOneClassWithOne α _ :=
+--   rfl
+
```
Comment (semorrison) [inline Mathlib/Algebra/Group/WithOne/Defs.lean:180]: Delete?

**Comment by semorrison** [comment]
If you can remember it, could you add a porting note to the proof you had to rewrite for lack of `simp`? It would be good, even if only as a test, to restore this when `lift` arrives, hopefully soon.

**Comment by semorrison** [comment]
bors d+

Please update the wiki page when you merge.

File: Mathlib/Algebra/Group/WithOne/Defs.lean [hunk 2]
```diff
@@ -105,10 +118,11 @@ theorem coe_unone {x : WithOne α} (hx : x ≠ 1) : ↑(unone hx) = x :=
...
 
-@[to_additive]
-theorem some_eq_coe {a : α} : (some a : WithOne α) = ↑a :=
-  rfl
-#align with_one.some_eq_coe WithOne.some_eq_coe
+-- porting note: in Lean 4 this is a syntactic tautology
+-- @[to_additive]
+-- theorem some_eq_coe {a : α} : (some a : WithOne α) = ↑a :=
+--   rfl
+-- #align with_one.some_eq_coe WithOne.some_eq_coe
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/WithOne/Defs.lean:125]:
```suggestion
#noalign with_one.some_eq_coe
```

File: Mathlib/Algebra/Group/WithOne/Defs.lean [hunk 3]
```diff
@@ -339,53 +373,59 @@ instance [DivisionMonoid α] : DivisionMonoid (WithZero α) :=
     inv_eq_of_mul := fun a b =>
       match a, b with
       | none, none => fun _ => rfl
-      | none, some b => by contradiction
-      | some a, none => by contradiction
+      | none, some b => λ _ =>  by contradiction
+      | some a, none => λ _ => by contradiction
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Group/WithOne/Defs.lean:377]: I think we use `fun` now

File: Mathlib/Algebra/Group/WithOne/Defs.lean (hunk 1 above)
Comment (kbuzzard) [reply Mathlib/Algebra/Group/WithOne/Defs.lean:180]: I'd have to make a corresponding PR to mathlib3 then, right? https://github.com/leanprover-community/mathlib/blob/c982179ec21091d3e102d8a5d9f5fe06c8fafb73/src/algebra/group/with_one/defs.lean#L127-L128

File: Mathlib/Algebra/Group/WithOne/Defs.lean (hunk 2 above)
Comment (kbuzzard) [reply Mathlib/Algebra/Group/WithOne/Defs.lean:125]: And `with_zero.some_eq_coe` too!

File: Mathlib/Algebra/Group/WithOne/Defs.lean (hunk 3 above)
Comment (kbuzzard) [reply Mathlib/Algebra/Group/WithOne/Defs.lean:377]: Thanks!

**Comment by kbuzzard** [comment]
OK I just realised that every `@[to_additive]` followed by an `#align` needs to have a second `#align` added for the additivised definition. I've just added these but don't want to hit merge until the branch compiles and unfortunately it's time for bed.

**Comment by kbuzzard** [comment]
Oh nice, it doesn't take hours to compile like mathlib3 :-)

**Comment by semorrison** [comment]
bors merge

File: Mathlib/Algebra/Group/WithOne/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    zpow_neg' := fun n x =>
+      match x with
+      | none => rfl
+      | some x => congr_arg some <| DivInvMonoid.zpow_neg' n x }
+
+instance [DivInvOneMonoid α] : DivInvOneMonoid (WithZero α) :=
+  { instDivInvMonoidWithZero, instInvOneClassWithZero with }
+
+instance [DivisionMonoid α] : DivisionMonoid (WithZero α) :=
+  { instDivInvMonoidWithZero, instHasInvolutiveInvWithZero with
```
Comment (semorrison) [inline Mathlib/Algebra/Group/WithOne/Defs.lean:382]:
```suggestion
  { instDivInvMonoidWithZero, instInvolutiveInvWithZero with
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**